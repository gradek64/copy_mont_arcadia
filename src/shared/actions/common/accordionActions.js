export function toggleAccordion(key, groupName = '') {
  return {
    type: 'TOGGLE_ACCORDION',
    key,
    groupName,
  }
}

export function openAccordion(key, groupName = '') {
  return {
    type: 'OPEN_ACCORDION',
    key,
    groupName,
  }
}

export function setInitiallyExpandedAccordion(key) {
  return {
    type: 'SET_INITIALLY_EXPANDED_ACCORDION',
    key,
  }
}

export function closeAllAccordions() {
  return {
    type: 'CLOSE_ALL_ACCORDIONS',
  }
}

export function closeAccordion(key) {
  return {
    type: 'CLOSE_ACCORDION',
    key,
  }
}

export function setScrolledIntoView(key, value) {
  return {
    type: 'SET_SCROLLED_INTO_VIEW',
    key,
    value,
  }
}

export function closeOtherAccordionsInGroup(groupName, accordionName) {
  return {
    type: 'CLOSE_OTHER_ACCORDIONS_IN_GROUP',
    groupName,
    accordionName,
  }
}
