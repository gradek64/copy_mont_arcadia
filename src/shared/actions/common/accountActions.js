import { updateMenuForAuthenticatedUser } from './navigationActions'
import { setFindAddressMode } from './checkoutActions'
import {
  updateOrderId,
  updateShoppingBagBadgeCount,
} from './shoppingBagActions'
import { get, post, put } from '../../lib/api-service'
import {
  getDefaultAddress,
  getDefaultPaymentOptions,
  isUserCreditCard,
} from '../../lib/checkout-utilities/actions-helpers'
import { getUnconfirmedDDPUser } from '../../lib/checkout-utilities/utils'
import { passwordMatchesPrevious } from '../../lib/account-utils'

import { path, pathOr, pluck } from 'ramda'
import { browserHistory } from 'react-router'
import {
  getUser,
  getExponeaMemberId,
  getExponeaLink,
} from '../../selectors/common/accountSelectors'
import { getFormNames, getAddressForm } from '../../selectors/formsSelectors'
import { getPaymentMethodTypeByValue } from '../../selectors/paymentMethodSelectors'
import { getRoutePath, isMyAccount } from '../../selectors/routingSelectors'
import { isUserAtLeastPartiallyAuthenticated } from '../../selectors/userAuthSelectors'
import { isFeatureMyPreferencesEnabled } from '../../selectors/featureSelectors'
import {
  setFormMeta,
  setFormMessage,
  resetForm,
  setFormSuccess,
  handleFormResponseErrorMessage,
  resetFormPartial,
  clearFormErrors,
} from './formActions'
import { ajaxCounter } from '../components/LoaderOverlayActions'
import {
  setPostResetUrl,
  changePasswordSuccess,
} from '../components/ChangePasswordActions'
import { setGenericError } from './errorMessageActions'
import { getAllPaymentMethods } from './paymentMethodsActions'
import { formNames } from '../../constants/forms'
import * as logger from '../../../server/lib/logger'
import * as clientLogger from '../../../client/lib/logger'
import { orderLinesSortedWithTracking } from '../../lib/order-details'
import { removeDiacriticsDeep } from '../../lib/checkout-utilities/klarna-utils'

export function userAccount(user) {
  return {
    type: 'USER_ACCOUNT',
    user,
  }
}

export const fetchExponeaLink = (link) => {
  return {
    type: 'FETCH_EXPONENA_LINK',
    link,
  }
}

export function closeCustomerDetailsModal() {
  return setFormMeta('customerDetails', 'modalOpen', false)
}

// Exponea action
export const getPreferenceLink = () => async (dispatch, getState) => {
  const { locale } = getState().config
  const memberId = getExponeaMemberId(getState())
  if (memberId) {
    try {
      const res = await dispatch(post('/exponea', { memberId }))
      const link = path(['body', 'link'], res)
      dispatch(fetchExponeaLink(`${link}?lang=${locale}`))
    } catch (error) {
      clientLogger.nrBrowserLogError('exponea api error', error)
    }
  }
}

export function getAccount() {
  return async (dispatch, getState) => {
    const state = getState()
    if (!isUserAtLeastPartiallyAuthenticated(state)) return
    dispatch(ajaxCounter('increment'))
    try {
      const response = await dispatch(get('/account'))

      // if user has bought ddp, review the end date of the subscription
      // because if the fraud check isn't complete the end date will be accurate
      // but isDDPUser status won't reflect the successful purchase
      const updatedUser = getUnconfirmedDDPUser(response.body)

      dispatch(userAccount(updatedUser))
      if (
        isFeatureMyPreferencesEnabled(state) &&
        !getExponeaLink(state) &&
        isMyAccount(state)
      ) {
        await dispatch(getPreferenceLink())
      }
    } catch (error) {
      dispatch(setGenericError(error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function updateAccount(data) {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))

    try {
      const response = await dispatch(put('/account/customerdetails', data))
      const formName = 'customerDetails'

      dispatch(setFormMeta(formName, 'modalOpen', true))
      dispatch(setFormSuccess(formName, true))
      dispatch(userAccount(response.body))
    } catch (error) {
      dispatch(setGenericError(error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function changeShortProfileRequest(data) {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))
    const formName = 'customerShortProfile'

    try {
      const response = await dispatch(
        put('/account/shortdetails', removeDiacriticsDeep(data))
      )
      dispatch(
        setFormMessage(formName, {
          type: 'confirm',
          message: 'Your profile details have been successfully updated.',
        })
      )
      dispatch(setFormSuccess(formName, true))
      dispatch(userAccount(response.body))
    } catch (error) {
      dispatch(handleFormResponseErrorMessage(formName, error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function changePwdRequest(data, resetPassword) {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))
    const formName = 'changePassword'

    try {
      const response = await dispatch(put('/account/changepassword', data))

      dispatch(userAccount(response.body))

      dispatch(
        setFormMessage(formName, {
          type: 'confirm',
          message: 'Your password has been successfully changed.',
        })
      )
      dispatch(setFormSuccess(formName, true))
      dispatch(changePasswordSuccess(true))

      if (resetPassword) {
        dispatch(getAccount())
      }
    } catch (error) {
      dispatch(handleFormResponseErrorMessage(formName, error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function setForgetPassword(value) {
  return {
    type: 'TOGGLE_FORGET_PASSWORD',
    value,
  }
}

export function toggleForgetPassword() {
  return {
    type: 'TOGGLE_FORGET_PASSWORD',
  }
}

function setResetUrl(state) {
  const pathname = getRoutePath(state)

  return setPostResetUrl(
    pathname.startsWith('/checkout') ? '/checkout' : '/my-account'
  )
}

export function forgetPwdRequest(data) {
  return async (dispatch, getState) => {
    dispatch(ajaxCounter('increment'))
    const formName = 'forgetPassword'

    try {
      const response = await dispatch(post('/account/forgetpassword', data))
      const message = path(['body', 'message'], response)
      dispatch(
        setFormMessage(formName, {
          type: 'confirm',
          message,
        })
      )
      dispatch(setFormSuccess(formName, true))
      dispatch(setResetUrl(getState()))
    } catch (error) {
      dispatch(handleFormResponseErrorMessage(formName, error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function resetPasswordLinkIsValid() {
  return {
    type: 'RESET_PASSWORD_LINK_VALID',
  }
}

export function resetPasswordLinkIsInvalid() {
  return {
    type: 'RESET_PASSWORD_LINK_INVALID',
  }
}
export function validateResetPasswordLinkExpiry(location) {
  return async (dispatch) => {
    const body = {
      storeId: location.query.storeId,
      email: location.query.token,
      hash: location.query.hash,
      catalogId: location.query.catalogId,
      langId: location.query.langId,
    }
    try {
      const response = await dispatch(
        post('/account/validate_reset_password', body)
      )
      if (response.body.isValidEmailLink) {
        dispatch(resetPasswordLinkIsValid())
      } else {
        dispatch(resetPasswordLinkIsInvalid())
      }
    } catch (e) {
      dispatch(resetPasswordLinkIsInvalid())
    }
  }
}

export function resetPasswordLinkRequest(data) {
  return async (dispatch, getState) => {
    dispatch(ajaxCounter('increment'))
    const formName = 'forgetPassword'

    try {
      const response = await dispatch(
        post('/account/reset_password_link', data)
      )
      const message = path(['body', 'message'], response)
      dispatch(
        setFormMessage(formName, {
          type: 'confirm',
          message,
        })
      )

      dispatch(setFormSuccess(formName, true))
      dispatch(setResetUrl(getState()))
    } catch (error) {
      dispatch(handleFormResponseErrorMessage(formName, error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function resetPasswordSuccess(authResponse) {
  return {
    type: 'RESET_PASSWORD_FORM_API_SUCCESS',
    payload: authResponse,
  }
}

export function resetPasswordRequest(data) {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))
    const formName = 'resetPassword'

    try {
      const authResponse = await dispatch(put('/account/reset_password', data))
      dispatch(
        setFormMessage(formName, {
          type: 'confirm',
          message: 'Your password has been successfully changed.',
        })
      )
      dispatch(setFormSuccess(formName, true))
      if (data.orderId) {
        dispatch(updateOrderId(data.orderId))
        dispatch(updateShoppingBagBadgeCount(authResponse.body.basketItemCount))
      }
      dispatch(userAccount(authResponse.body))
      dispatch(resetPasswordSuccess(authResponse.body))
      dispatch(updateMenuForAuthenticatedUser())
    } catch (error) {
      const errorMessage = pathOr('', ['response', 'body', 'message'], error)
      const errorCode = path(['response', 'body', 'errorCode'], error)

      if (passwordMatchesPrevious(errorCode)) {
        dispatch(handleFormResponseErrorMessage(formName, error, errorMessage))
      } else {
        dispatch(resetPasswordLinkIsInvalid())
      }
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

const ORDER_HISTORY = '/account/order-history'

export function setOrderHistoryDetails(orderDetails) {
  const { orderLines } = orderDetails
  const orderLinesSortedByTracking = orderLinesSortedWithTracking(orderLines)
  return {
    type: 'SET_ORDER_HISTORY_DETAILS',
    orderDetails: { ...orderDetails, orderLinesSortedByTracking },
  }
}

export function setOrderHistoryOrders(orders) {
  return {
    type: 'SET_ORDER_HISTORY_ORDERS',
    orders,
  }
}

export function orderHistoryRequest() {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))

    try {
      const response = await dispatch(get(ORDER_HISTORY))

      dispatch(setOrderHistoryOrders(path(['body', 'orders'], response)))
    } catch (error) {
      dispatch(setOrderHistoryOrders({}))
      dispatch(setGenericError(error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function orderHistoryDetailsRequest(orderId) {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))

    try {
      const response = await dispatch(get(`${ORDER_HISTORY}/${orderId}`))

      dispatch(setOrderHistoryDetails(response.body))
    } catch (error) {
      dispatch(setGenericError(error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

const RETURN_HISTORY = '/account/return-history'

export function setReturnHistoryReturns(returns) {
  return {
    type: 'SET_RETURN_HISTORY_RETURNS',
    returns,
  }
}

export function returnHistoryRequest() {
  return async (dispatch) => {
    dispatch(ajaxCounter('increment'))

    try {
      const response = await dispatch(get(RETURN_HISTORY))

      dispatch(setReturnHistoryReturns(path(['body', 'orders'], response)))
    } catch (error) {
      dispatch(setGenericError(error))
    } finally {
      dispatch(ajaxCounter('decrement'))
    }
  }
}

export function setReturnHistoryDetails(returnDetails) {
  return {
    type: 'SET_RETURN_HISTORY_DETAILS',
    returnDetails,
  }
}

export const returnHistoryDetailsRequest = (orderId, rmaId) => async (
  dispatch
) => {
  try {
    dispatch(ajaxCounter('increment'))
    const { body: returnHistoryDetails } = await dispatch(
      get(`${RETURN_HISTORY}/${orderId}/${rmaId}`)
    )
    dispatch(setReturnHistoryDetails(returnHistoryDetails))
    dispatch(ajaxCounter('decrement'))
  } catch (e) {
    logger.info('server-side-render', 'returnHistoryDetailsRequest', e)
    dispatch(setReturnHistoryDetails({}))
    dispatch(ajaxCounter('decrement'))
  }
}

/**
 * My Checkout Details actions
 */
export function setMyCheckoutDetailsInitialFocus(initialFocus) {
  return {
    type: 'SET_MCD_INITIAL_FOCUS',
    initialFocus,
  }
}

// fill the forms with account.user (Redux) data
export function resetMyCheckoutDetailsForms() {
  return (dispatch, getState) => {
    const state = getState()
    const user = getUser(state)
    const { siteOptions } = state
    const deliveryCountry = pathOr(
      null,
      ['deliveryDetails', 'address', 'country'],
      user
    )
    const defaultState =
      deliveryCountry === 'United States' ? siteOptions.USStates[0] : null

    const defaultAddress = {
      ...getDefaultAddress(),
      country: deliveryCountry,
      state: defaultState,
    }

    const deliveryAddress = pathOr(
      defaultAddress,
      ['deliveryDetails', 'address'],
      user
    )

    dispatch(
      resetForm(formNames.deliveryMCD.address, {
        ...deliveryAddress,
        county: null,
        isManual: true,
      })
    )

    const deliveryTelephone = pathOr(
      null,
      ['deliveryDetails', 'nameAndPhone', 'telephone'],
      user
    )
    const deliveryNameAndPhone = path(['deliveryDetails', 'nameAndPhone'], user)
    dispatch(
      resetForm(formNames.deliveryMCD.details, {
        ...deliveryNameAndPhone,
        telephone: deliveryTelephone,
      })
    )

    const billingAddress = pathOr(
      defaultAddress,
      ['billingDetails', 'address'],
      user
    )
    dispatch(
      resetForm(formNames.billingMCD.address, {
        ...billingAddress,
        county: null,
        isManual: true,
      })
    )

    const billingTelephone = pathOr(
      null,
      ['billingDetails', 'nameAndPhone', 'telephone'],
      user
    )
    const billingNameAndPhone = path(['billingDetails', 'nameAndPhone'], user)
    dispatch(
      resetForm(formNames.billingMCD.details, {
        ...billingNameAndPhone,
        telephone: billingTelephone,
      })
    )

    dispatch(setFindAddressMode())
    // Payment details
    const defaultPaymentOptions = getDefaultPaymentOptions()
    const { type: value, expiryMonth, expiryYear } = pathOr(
      {},
      ['creditCard'],
      user
    )
    const paymentMethodTypeForUser = getPaymentMethodTypeByValue(state, value)

    const setExpiryFor = (monthOrYear) => {
      switch (monthOrYear) {
        case 'month':
          return paymentMethodTypeForUser === 'OTHER'
            ? defaultPaymentOptions.expiryMonth
            : expiryMonth

        case 'year':
          return paymentMethodTypeForUser === 'OTHER'
            ? defaultPaymentOptions.expiryYear
            : expiryYear

        default:
          return ''
      }
    }

    const convertToExpireDateFormat = () => {
      const expiryMonthStr = setExpiryFor('month')
      const expiryYearStr = setExpiryFor('year')

      return `${expiryMonthStr}${expiryYearStr.length > 0 &&
        expiryYearStr.substr(2)}`
    }

    const userCardDetails = isUserCreditCard(user)
      ? {
          ...defaultPaymentOptions,
          paymentType: value,
          expiryDate: convertToExpireDateFormat(),
          expiryMonth: setExpiryFor('month'),
          expiryYear: setExpiryFor('year'),
        }
      : null

    dispatch(
      resetForm(
        formNames.payment.paymentCardDetailsMCD,
        userCardDetails || defaultPaymentOptions
      )
    )
  }
}

/**
 * Action used on MyChekoutDetails to grab user data and fill MyCheckoutDetails forms
 */
export function getMyCheckoutDetailsData() {
  return (dispatch) => {
    dispatch(ajaxCounter('increment'))
    return dispatch(getAccount()) // TODO double fetch of account (this + MyAccountSubcategory needs)
      .then(() => {
        dispatch(ajaxCounter('decrement'))
        dispatch(resetMyCheckoutDetailsForms())
        // ADP-3321: Replacing getPaymentMethods with getAllPaymentMethods.
        // This is a provisional change as the ultimate goal is populate paymentMethods in the SSR.
        // @TODO To be removed in ADP-3653
        return dispatch(getAllPaymentMethods())
      })
      .catch((err) => {
        dispatch(ajaxCounter('decrement'))
        dispatch(
          handleFormResponseErrorMessage(
            'myCheckoutDetailsForm',
            err,
            'An error has occurred. Please try again.'
          )
        )
      })
  }
}

/**
 * Action that copies form data from billing forms to delivery forms
 * @param {object} billingFormNames object with one key per each form name
 * @param {object} deliveryFormNames object with one key per each form name
 */
export function setDeliveryFormsFromBillingForms() {
  return (dispatch, getState) => {
    const state = getState()
    const billingFormNames = getFormNames('billingMCD')
    const deliveryFormNames = getFormNames('deliveryMCD')

    // billingFormNames and deliveryFormNames objects should have the same keys
    Object.keys(billingFormNames).forEach((formKey) => {
      const billingFormName = billingFormNames[formKey]
      const deliveryFormName = deliveryFormNames[formKey]
      const form = getAddressForm('deliveryMCD', billingFormName, state)
      const newFields = pluck('value', form.fields)
      dispatch(resetForm(deliveryFormName, newFields))
      // Cleaning up any errors in the delivery form when copied from billing.
      // This is to prevent any existing errors from persisting
      // This isn't an ideal strategy but good enough until we refactor the
      // forms and perhaps this entire page.
      dispatch(clearFormErrors(deliveryFormName))
    })
  }
}

/**
 * Action that update user information using the My Checkout Details form
 * @param {*} data object
 * @param {*} formName form name where to write messages (confirm or error)
 */
export function updateMyCheckoutDetails(data, formName) {
  return (dispatch) => {
    dispatch(ajaxCounter('increment'))
    return dispatch(put('/account/customerdetails', data))
      .then((res) => {
        dispatch(userAccount(res.body))
        dispatch(ajaxCounter('decrement'))
        dispatch(
          setFormMessage(formName, {
            type: 'confirm',
            message: 'Your changes have been saved',
          })
        )

        dispatch(resetFormPartial('paymentCardDetailsMCD', { cardNumber: '' }))
        browserHistory.push('/my-account/details')
      })
      .catch((err) => {
        dispatch(ajaxCounter('decrement'))
        dispatch(
          handleFormResponseErrorMessage(
            formName,
            err,
            `An error has occurred. Please try again.`
          )
        )
      })
  }
}

export function checkAccountExists({
  email = '',
  successCallback,
  failureCallback,
}) {
  return async (dispatch) => {
    try {
      const res = await dispatch(get(`/account?email=${email}`))
      if (successCallback) successCallback(res)
    } catch (e) {
      logger.error(e)
      if (failureCallback) failureCallback(e)
    }
  }
}
