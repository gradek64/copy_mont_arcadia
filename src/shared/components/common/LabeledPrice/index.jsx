import React from 'react'
import PropTypes from 'prop-types'
import cn from 'classnames'

import Price from '../Price/Price'

const LabeledPrice = ({
  rrp,
  wasWasPrice,
  wasPrice,
  regPrice,
  brandName,
  l,
}) => {
  const price = rrp || wasWasPrice || wasPrice || regPrice
  let label = ''
  if (rrp) label = 'RRP'
  if (wasWasPrice || wasPrice) label = 'Was'
  if (regPrice) label = 'Now'
  if (brandName === 'dorothyperkins' || brandName === 'burton') label = ''

  return (
    <span
      className={cn({
        'HistoricalPrice-old': !regPrice,
        'HistoricalPrice-rrp': rrp,
        'HistoricalPrice-wasWas': wasWasPrice,
        'HistoricalPrice-was': wasPrice,
        'HistoricalPrice-promotion': regPrice,
      })}
    >
      {/* eslint-disable-next-line  */}
      {l`${label}`} <Price price={price} />
    </span>
  )
}

LabeledPrice.propTypes = {
  rrp: PropTypes.string,
  wasWasPrice: PropTypes.string,
  wasPrice: PropTypes.string,
  regPrice: PropTypes.string,
  brandName: PropTypes.string.isRequired,
  l: PropTypes.func.isRequired,
}

export default LabeledPrice
