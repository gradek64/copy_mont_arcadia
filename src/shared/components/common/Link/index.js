import React from 'react'
import PropTypes from 'prop-types'
import { Link as RRLink } from 'react-router'

/* eslint-disable jsx-a11y/anchor-has-content */
export default function Link({
  isExternal = false,
  to,
  onlyActiveOnIndex = false,
  ...restProps
}) {
  if (isExternal) return <a {...restProps} href={to} />

  return <RRLink {...restProps} to={to} onlyActiveOnIndex={onlyActiveOnIndex} />
}

Link.propTypes = {
  ...RRLink.propTypes,
  children: PropTypes.node.isRequired,
  isExternal: PropTypes.bool,
  onlyActiveOnIndex: PropTypes.bool,
}
