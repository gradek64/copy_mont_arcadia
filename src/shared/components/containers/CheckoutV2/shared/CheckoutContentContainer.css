/** @define CheckoutContentContainer; weak */
.CheckoutContentContainer {
  $border: 1px solid $accordion-border-color;
  margin: $checkout-container-margin-vertical $page-padding;
  padding-top: $checkout-separator-height;
  &-content {
    padding-top: 0;
  }
}

@media $min-tablet {
  .CheckoutContentContainer {
    padding-top: 1px;
  }
}
