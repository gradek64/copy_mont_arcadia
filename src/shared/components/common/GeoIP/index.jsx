import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { pick } from 'ramda'

import url from 'url'
import { getRedirectURL, getURLPath } from '../../../lib/geo-ip-utils'
import { closeModal } from '../../../actions/common/modalActions'
import { sendAnalyticsDisplayEvent } from '../../../../shared/analytics/analytics-actions'
import { error, info } from '../../../../client/lib/logger'
import RedirectionPrompt from './RedirectionPrompt'
import GeoIPPixelCookies from './GeoIPPixelCookies'
import { GTM_EVENT } from '../../../analytics'
import { toggleLoaderOverlay } from '../../../actions/components/LoaderOverlayActions'
import {
  getGeoIPUserISOPreference,
  getGeoIPUserRegionPreference,
  getGeoIPUserLanguagePreference,
} from '../../../selectors/geoIPSelectors'
import { getLang, getLanguageRegion } from '../../../selectors/configSelectors'

import { translateGeoIPTextInPreferredLanguage } from '../../../lib/localisation'

export const addClientURLPartsToLink = (link) => {
  const { port = null, protocol = null } = window.location
  link = protocol ? `${protocol}//${link}` : link

  const parsedLink = url.parse(link)
  delete parsedLink.host
  delete parsedLink.href
  parsedLink.port = port
  return url.format(parsedLink)
}

@connect(
  (state) => {
    const redirectURL = getRedirectURL(state)
    const showFootnote =
      redirectURL && /^\/(\?|#|$)/.test(getURLPath(redirectURL))

    return {
      redirectURL,
      currentSiteRegion: state.config.region,
      currentSiteISO: state.config.preferredISOs[0],
      userISOPreference: getGeoIPUserISOPreference(state),
      userRegionPreference: getGeoIPUserRegionPreference(state),
      userLanguagePreference: getGeoIPUserLanguagePreference(state),
      brandName: state.config.brandName,
      showFootnote,
      geoIPReduxState: state.geoIP,
      currentSiteLanguage: getLang(state),
      currentSiteLanguageRegion: getLanguageRegion(state),
    }
  },
  { closeModal, sendAnalyticsDisplayEvent, toggleLoaderOverlay }
)
class GeoIPModal extends React.Component {
  static propTypes = {
    redirectURL: PropTypes.string.isRequired,
    userISOPreference: PropTypes.string.isRequired,
    userRegionPreference: PropTypes.string,
    userLanguagePreference: PropTypes.string,
    handleClose: PropTypes.func.isRequired,
    closeModal: PropTypes.func.isRequired,
    showFootnote: PropTypes.bool.isRequired,
    geoIPReduxState: PropTypes.object.isRequired,
    currentSiteISO: PropTypes.string.isRequired,
    currentSiteRegion: PropTypes.string,
    currentSiteLanguage: PropTypes.string,
    currentSiteLanguageRegion: PropTypes.string,
    handleRemoveCloseHandler: PropTypes.func,
    sendAnalyticsDisplayEvent: PropTypes.func.isRequired,
    toggleLoaderOverlay: PropTypes.func.isRequired,
  }

  static defaultProps = {
    currentSiteRegion: 'uk',
    currentSiteLanguage: 'en',
    currentSiteLanguageRegion: 'en-gb',
    userRegionPreference: 'uk',
    userLanguagePreference: 'en-gb',
  }

  constructor(props) {
    super(props)

    this.state = {
      cookieValue: null,
      shouldRedirect: false,
      shouldSetGeoIPCookies: false,
    }

    this.geoTranslate = translateGeoIPTextInPreferredLanguage.bind(
      null,
      this.props.userLanguagePreference,
      this.props.currentSiteLanguageRegion,
      this.props.brandName
    )
  }

  /**
   * Logic is here in the componentWillMount in order pass the closer handler to
   * the parent Modal component as well as the separate ContentOverlay component;
   * following the observer pattern. We do this as it is bad practice and an anti-pattern
   * to store functions in the redux state, coupled with the way the Modal and
   * ContentOverlay components have been written and separated out. Ultimately
   * we should look into rewriting the Modal logic to accommodate this new use case:
   * requiring some extra logic to be handled upon closing the Modal.
   */
  UNSAFE_componentWillMount() {
    this.props.handleClose(() => {
      if (this.finishedGeoPixel)
        throw new Error('Cannot call handleClose callback multiple times')

      return new Promise((resolve) => {
        this.finishedGeoPixel = resolve
        this.setGeoIPCookies({
          cookieValue: this.props.currentSiteISO,
          shouldRedirect: false,
        })()
      })
    })
  }

  componentDidMount() {
    const {
      sendAnalyticsDisplayEvent,
      userRegionPreference,
      currentSiteRegion,
    } = this.props

    sendAnalyticsDisplayEvent(
      {
        currentLocale: currentSiteRegion,
        suggestedLocale: userRegionPreference,
      },
      GTM_EVENT.GEO_IP_MODAL_DISPLAYED
    )
  }

  componentWillUnmount() {
    this.props.handleRemoveCloseHandler()
  }

  setGeoIPCookies = ({ cookieValue, shouldRedirect }) => () => {
    const { toggleLoaderOverlay } = this.props
    toggleLoaderOverlay()
    this.setState({
      shouldSetGeoIPCookies: true,
      shouldRedirect,
      cookieValue,
    })
  }

  handleGeoIPCookieError = (err) => {
    const { geoIPReduxState } = this.props
    const { shouldRedirect, cookieValue } = this.state

    error('GeoIP', {
      message: 'GeoIP pixel error when setting cookie for all hostnames',
      sourceURL: window.location.href,
      geoPixelURL: err.nativeEvent.currentTarget.currentSrc,
      newGeoPreference: cookieValue,
      userRequestedRedirect: shouldRedirect,
      geoIPReduxState,
    })
  }

  handleGeoIPCookieSuccess = () => {
    const { shouldRedirect, cookieValue } = this.state
    const {
      redirectURL,
      geoIPReduxState,
      currentSiteISO,
      closeModal,
      toggleLoaderOverlay,
    } = this.props

    if (shouldRedirect) {
      info('GeoIP', {
        message: 'User has been redirected to their preferred site',
        sourceURL: window.location.href,
        newGeoPreference: cookieValue,
        geoIPReduxState,
        redirectURL,
      })
      return window.location.assign(addClientURLPartsToLink(redirectURL))
    }
    info('GeoIP', {
      message: 'User has chosen to stay on the requested site',
      sourceURL: window.location.href,
      newGeoPreference: currentSiteISO,
      geoIPReduxState,
      redirectURL,
    })
    if (this.finishedGeoPixel) {
      this.finishedGeoPixel()
    } else {
      closeModal()
    }
    toggleLoaderOverlay()
  }

  getRedirectionPromptProps() {
    return {
      ...pick(
        [
          'currentSiteRegion',
          'userRegionPreference',
          'userISOPreference',
          'currentSiteISO',
          'showFootnote',
        ],
        this.props
      ),
      setGeoIPCookies: this.setGeoIPCookies,
      geoTranslate: this.geoTranslate,
    }
  }

  render() {
    /**
     * We translate the text in the GeoIPModal based on the userISOPreference,
     * rather than using the default site language and so we have to maintain
     * a mini dictionary for it (see translateGeoIPTextInPreferredLanguage).
     *
     * If you change or add a subcomponent please make sure all strings are
     * included in geoIPDictionaryStrings - see shared/lib/localisation.js
     */

    return (
      <div className="GeoIP" data-id="GeoIPModal">
        <div>
          <RedirectionPrompt {...this.getRedirectionPromptProps()} />
          {this.state.shouldSetGeoIPCookies && (
            <GeoIPPixelCookies
              handleSuccess={this.handleGeoIPCookieSuccess}
              handleError={this.handleGeoIPCookieError}
              cookieValue={this.state.cookieValue}
            />
          )}
        </div>
      </div>
    )
  }
}

export default GeoIPModal
