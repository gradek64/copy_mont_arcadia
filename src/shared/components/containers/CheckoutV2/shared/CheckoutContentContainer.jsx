import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { openMiniBag } from '../../../../actions/common/shoppingBagActions'
import { isMobile } from '../../../../selectors/viewportSelectors'
import { getOrderCost } from '../../../../selectors/checkoutSelectors'

import QuickViewOrderSummary from '../../../common/QuickViewOrderSummary/QuickViewOrderSummary'

const mapDispatchToProps = { openMiniBag }

@connect(
  (state) => ({
    totalCost: getOrderCost(state),
    isMobile: isMobile(state),
  }),
  mapDispatchToProps
)
class CheckoutContentContainer extends Component {
  static contextTypes = {
    l: PropTypes.func,
    p: PropTypes.func,
  }

  componentDidMount() {
    if (window) window.scrollTo(0, 0)
  }

  render() {
    const { children, isMobile, totalCost, openMiniBag } = this.props
    const { l, p } = this.context

    return (
      <section className="CheckoutContentContainer">
        <QuickViewOrderSummary
          isMobile={isMobile}
          openMiniBag={openMiniBag}
          totalCost={p(totalCost)}
          l={l}
        />
        <div className="CheckoutContentContainer-content">{children}</div>
      </section>
    )
  }
}

export default CheckoutContentContainer
