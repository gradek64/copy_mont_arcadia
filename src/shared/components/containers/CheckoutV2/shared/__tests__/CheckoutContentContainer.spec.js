import { openMiniBag } from '../../../../../actions/common/shoppingBagActions'
import testComponent, {
  renderConnectedComponentProps,
} from '../../../../../../../test/unit/helpers/test-component'
import CheckoutContentContainer from '../CheckoutContentContainer'

jest.mock('../../../../../actions/common/shoppingBagActions', () => ({
  openMiniBag: jest.fn(() => ({ type: 'foo' })),
}))

const renderComponent = testComponent(CheckoutContentContainer.WrappedComponent)

describe('@renders', () => {
  describe('the default rendering', () => {
    const props = {
      getAccount: jest.fn(),
    }
    const component = renderComponent(props)

    it('renders as expected', () => {
      expect(component.getTree()).toMatchSnapshot()
    })
  })
})

describe('@connected', () => {
  describe('mapping dispatch to props', () => {
    const props = renderConnectedComponentProps(CheckoutContentContainer)
    props.openMiniBag('bar')

    it('correctly maps dispatch to props', () => {
      expect(openMiniBag).toHaveBeenCalledTimes(1)
      expect(openMiniBag).toHaveBeenCalledWith('bar')
    })
  })
})
