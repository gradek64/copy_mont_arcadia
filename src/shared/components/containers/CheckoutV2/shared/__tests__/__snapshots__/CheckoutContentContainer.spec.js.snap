// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@renders the default rendering renders as expected 1`] = `
<section
  className="CheckoutContentContainer"
>
  <QuickViewOrderSummary
    l={[MockFunction]}
    totalCost="Â£NaN"
  />
  <div
    className="CheckoutContentContainer-content"
  />
</section>
`;
